File: java/components/validation-ctt/HEAD/ctt_testcases_v14.12.00b/src/com/at4wireless/alljoyn/core/notification/NotificationValidator.java
/*
 * Copyright AllSeen Alliance. All rights reserved.
 *
 *    Permission to use, copy, modify, and/or distribute this software for any
 *    purpose with or without fee is hereby granted, provided that the above
 *    copyright notice and this permission notice appear in all copies.
 *
 *    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 *    WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 *    MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 *    ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 *    WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 *    ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 *    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */
package com.at4wireless.alljoyn.core.notification;

import java.net.MalformedURLException;
import java.net.URL;
import java.util.List;
import java.util.UUID;
import java.util.concurrent.atomic.AtomicInteger;

import org.alljoyn.bus.ErrorReplyBusException;
import org.alljoyn.ns.Notification;
import org.alljoyn.ns.NotificationMessageType;
import org.alljoyn.ns.NotificationReceiver;
import org.alljoyn.ns.RichAudioUrl;

import com.at4wireless.alljoyn.core.about.AboutAnnouncementDetails;
import com.at4wireless.alljoyn.core.commons.log.WindowsLoggerImpl;
import com.at4wireless.alljoyn.core.introspection.BusIntrospector;



// TODO: Auto-generated Javadoc
/**
 * The Class NotificationValidator.
 */
public class NotificationValidator implements NotificationReceiver, Runnable
{
	
    
    /** The IXIT n_ notification version. */
    private int IXITN_NotificationVersion;
    
    /** The ICS n_ rich icon url. */
    private boolean ICSN_RichIconUrl;
    
    /** The ICS n_ rich audio url. */
    private boolean  ICSN_RichAudioUrl;
    
    /** The ICS n_ resp object path. */
    private boolean  ICSN_RespObjectPath;
    
    /** The Constant TAG. */
    private static final String TAG = "NotificationValidator";
	
	/** The Constant logger. */
	private static final WindowsLoggerImpl logger =  new WindowsLoggerImpl(TAG);
    
    /** The notification counter. */
    private AtomicInteger notificationCounter = new AtomicInteger();
    
    /** The bus introspector. */
    private BusIntrospector busIntrospector;
    
    /** The device id. */
    String deviceId;
    
    /** The app id. */
    UUID appId;
    
    /** The notification validation exception handler. */
    private NotificationValidationExceptionHandler notificationValidationExceptionHandler;
    
    /** The device about announcement. */
    private AboutAnnouncementDetails deviceAboutAnnouncement;
    
    /** The notification handler. */
    private NotificationHandler notificationHandler;

    /**
     * The Interface NotificationValidationExceptionHandler.
     */
    public interface NotificationValidationExceptionHandler
    {
        
        /**
         * On notification validation exception.
         *
         * @param exception the exception
         */
        void onNotificationValidationException(Exception exception);
    }

    /**
     * Instantiates a new notification validator.
     *
     * @param dutDeviceId the dut device id
     * @param dutAppIddeviceId the dut app iddevice id
     */
    public NotificationValidator(String dutDeviceId, UUID dutAppIddeviceId)
    {
    	deviceId=dutDeviceId;
    			appId=	dutAppIddeviceId;
        notificationHandler = getNotificationHandler();
        notificationHandler.initializeForDevice(dutDeviceId, dutAppIddeviceId);
   
    }

    /**
     * Gets the notification handler.
     *
     * @return the notification handler
     */
    protected NotificationHandler getNotificationHandler()
    {
        return new NotificationHandler();
    }

    /**
     * Initialize for device.
     *
     * @param deviceAboutAnnouncement the device about announcement
     * @param busIntrospector the bus introspector
     * @param notificationValidationExceptionHandler the notification validation exception handler
     */
    public void initializeForDevice(AboutAnnouncementDetails deviceAboutAnnouncement, BusIntrospector busIntrospector,
            NotificationValidationExceptionHandler notificationValidationExceptionHandler)
    {
        this.deviceAboutAnnouncement = deviceAboutAnnouncement;
        this.busIntrospector = busIntrospector;
        this.notificationValidationExceptionHandler = notificationValidationExceptionHandler;
    }

    /* (non-Javadoc)
     * @see org.alljoyn.ns.NotificationReceiver#receive(org.alljoyn.ns.Notification)
     */
    @Override
    public void receive(Notification notification)
    {
        notificationHandler.receive(notification);
    }

    /* (non-Javadoc)
     * @see org.alljoyn.ns.NotificationReceiver#dismiss(int, java.util.UUID)
     */
    @Override
    public void dismiss(int arg0, UUID arg1)
    {
    }

    /* (non-Javadoc)
     * @see java.lang.Runnable#run()
     */
    @Override
    public void run()
    {
        Notification notification;

        try
        {
            while ((notification = notificationHandler.getReceivedNotification()) != null)
            {
                logger.info(String.format("Validating notification from DUT with messageId: %d", notification.getMessageId()));

                try
                {
                	 notificationCounter.getAndIncrement();
                    boolean includedRichIconUrl = false;
                    boolean includedRichAudioUrl = false;
                    boolean includesResponseObjectPath = false;

                    assertEquals(String.format("Notification version received: "+notification.getVersion()+" does not match IXIT: "
                    		+IXITN_NotificationVersion), IXITN_NotificationVersion, notification.getVersion());
                    logger.info("Notification version received: "+notification.getVersion()+" matches IXIT: "
                    		+IXITN_NotificationVersion);
                    logger.info("Partial Verdict: PASS");
                    //String notifAppName = notification.getAppName();
                    assertEquals("AppName parameter received in notification: "+notification.getAppName()+" does not match About Announcement parameter: "
                    		+deviceAboutAnnouncement.getAppName(), deviceAboutAnnouncement.getAppName(), notification.getAppName());
                    logger.info("AppName parameter received: "+notification.getAppName()+" matches About Announcement parameter: "
                    		+deviceAboutAnnouncement.getAppName());
                    logger.info("Partial Verdict: PASS");
                    //String notifdeviceName = notification.getDeviceName();
                    assertEquals("DeviceName parameter received in notification: "+notification.getDeviceName()+" does not match About Announcement parameter: "
                    		+deviceAboutAnnouncement.getDeviceName(), deviceAboutAnnouncement.getDeviceName(), notification.getDeviceName());
                    logger.info("DeviceName parameter received in notification: "+notification.getDeviceName()+" matches About Announcement parameter: "
                    		+deviceAboutAnnouncement.getDeviceName());
                    logger.info("Partial Verdict: PASS");
                    assertNotNull("Notification messageType value not valid!", notification.getMessageType());
                    logger.info("Valid notification messageType value: "+notification.getMessageType());
                    logger.info("Partial Verdict: PASS");
                    logger.info("The received Message Id: " + notification.getMessageId());

                    String richIconUrlString = notification.getRichIconUrl();

                    if (ICSN_RichIconUrl && richIconUrlString != null)
                    {
                        try
                        {
                            logger.info("Rich icon URL: '" + richIconUrlString + "'");
                            new URL(richIconUrlString);
                            includedRichIconUrl = true;
                        }
                        catch (MalformedURLException e)
                        {
                            throw new Exception("RichIconUrl malformed: '" + richIconUrlString + "'", e);
                        }
                    }

                    List<RichAudioUrl> richAudioUrls = notification.getRichAudioUrl();

                    if (ICSN_RichAudioUrl && richAudioUrls != null)
                    {
                        for (RichAudioUrl richAudioUrl : richAudioUrls)
                        {
                            String urlString = richAudioUrl.getUrl();
                            try
                            {
                                new URL(urlString);
                                logger.info("Rich audio URL: '" + urlString + "'");
                                includedRichAudioUrl = true;
                            }
                            catch (MalformedURLException e)
                            {
                                throw new Exception("RichAudioUrl malformed: '" + urlString + "'", e);
                            }
                        }
                    }

                    String responseObjectPath = notification.getResponseObjectPath();
                    if (ICSN_RespObjectPath && responseObjectPath != null)
                    {
                        logger.info("Control Panel Service response Object Path: '" + responseObjectPath + "'");

                        try
                        {
                            busIntrospector.introspect(responseObjectPath);
                        }
                        catch (ErrorReplyBusException e)
                        {
                            throw new Exception("Exception retrieving responseObject", e);
                        }

                        includesResponseObjectPath = true;
                    }
                   

                    StringBuilder noteMsg = new StringBuilder();
                    noteMsg.append("Notification Received: ");
                    noteMsg.append("messageType: ");
                    noteMsg.append(notification.getMessageType());
                    noteMsg.append("; richIconUrl: ");
                    noteMsg.append(includedRichIconUrl);
                    noteMsg.append("; richAudioUrl: ");
                    noteMsg.append(includedRichAudioUrl);
                    noteMsg.append("; responseObject: ");
                    noteMsg.append(includesResponseObjectPath);
                    String noteMsgString = noteMsg.toString();
                    logger.addNote(noteMsgString);
                    logger.info(noteMsgString);
                }
                catch (Throwable throwable)
                {
                    if (throwable instanceof Exception)
                    {
                        notificationValidationExceptionHandler.onNotificationValidationException((Exception) throwable);
                    }
                    else
                    {
                        notificationValidationExceptionHandler.onNotificationValidationException(new Exception(throwable));
                    }
                }
            }
        }
        catch (InterruptedException e)
        {
            Thread.currentThread().interrupt();
        }
    }
    
    

    /**
     * Assert not null.
     *
     * @param errorMsg the error msg
     * @param messageType the message type
     * @throws Exception the exception
     */
    private void assertNotNull(String errorMsg,
			NotificationMessageType messageType) throws Exception {
		// TODO Auto-generated method stub
    	if(messageType==null)
    	{
    		logger.error(errorMsg);
    		throw new Exception(errorMsg);
    	}
    	
		
	}

	/**
	 * Assert equals.
	 *
	 * @param errorMsg the error msg
	 * @param appName the app name
	 * @param notifAppName the notif app name
	 * @throws Exception the exception
	 */
	private void assertEquals(String errorMsg, String appName, String notifAppName) throws Exception {
		// TODO Auto-generated method stub
		if(!appName.equals(notifAppName)){
			
			logger.error(errorMsg);
			throw new Exception(errorMsg);
			
		} else {
			logger.info("Partial Verdict: PASS");
		}
	}

	/**
	 * Assert equals.
	 *
	 * @param errorMsg the error msg
	 * @param notificationServiceVersion the notification service version
	 * @param version the version
	 * @throws Exception the exception
	 */
	private void assertEquals(String errorMsg, int notificationServiceVersion,
			int version) throws Exception {
		// TODO Auto-generated method stub
		
		if(notificationServiceVersion!=version){
		logger.error(errorMsg);
		throw new Exception(errorMsg);
		} else {
			logger.info("Partial Verdict: PASS");
		}
	}

	/**
	 * Gets the number of notifications received.
	 *
	 * @return the number of notifications received
	 */
	public int getNumberOfNotificationsReceived()
    {
        return notificationCounter.get();
    }

	/**
	 * Sets the test parameters.
	 *
	 * @param iCSN_RichIconUrl the i cs n_ rich icon url
	 * @param iCSN_RichAudioUrl the i cs n_ rich audio url
	 * @param iCSN_RespObjectPath the i cs n_ resp object path
	 * @param iXITN_NotificationVersion the i xit n_ notification version
	 */
	public void setTestParameters(boolean iCSN_RichIconUrl ,boolean iCSN_RichAudioUrl,boolean iCSN_RespObjectPath, String iXITN_NotificationVersion) {
		// TODO Auto-generated method stub
		
		ICSN_RichIconUrl=iCSN_RichIconUrl;
		ICSN_RichAudioUrl=iCSN_RichAudioUrl;
		ICSN_RespObjectPath=iCSN_RespObjectPath;
		
		IXITN_NotificationVersion=Integer.parseInt(iXITN_NotificationVersion);
		
	}

	
}

File: cpp/components/validation-ctt/HEAD/ctt_testcases/NotificationValidator.cpp
/******************************************************************************
* Copyright AllSeen Alliance. All rights reserved.
*
*    Permission to use, copy, modify, and/or distribute this software for any
*    purpose with or without fee is hereby granted, provided that the above
*    copyright notice and this permission notice appear in all copies.
*
*    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
*    WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
*    MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
*    ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
*    WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
*    ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
*    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
******************************************************************************/
#include "stdafx.h"
#include "NotificationValidator.h"

#include <regex>

#include "MessageSet.h"

const char* NotificationValidator::URL_REGEX = R"(^(([^:\/?#]+):)?(//([^\/?#]*))?([^?#]*)(\?([^#]*))?(#(.*))?)";

NotificationValidator::NotificationValidator(std::string t_DutDeviceId, uint8_t* t_DutAppId,
	bool t_ExpectedRichIconUrl, bool t_ExpectedRichAudioUrl, bool t_ExpectedResponseObjectPath,
	uint16_t t_ExpectedNotificationVersion) : 
	m_ExpectedRichIconUrl(t_ExpectedRichIconUrl),
	m_ExpectedRichAudioUrl(t_ExpectedRichAudioUrl), 
	m_ExpectedResponseObjectPath(t_ExpectedResponseObjectPath),
	m_ExpectedNotificationVersion(t_ExpectedNotificationVersion),
	m_ReadingNotifications(false)
{
	m_NotificationHandler = new NotificationHandler(t_DutDeviceId, t_DutAppId);
}

void NotificationValidator::initializeForDevice(AboutAnnouncementDetails& t_DeviceAboutAnnouncement,
	 XMLBasedBusIntrospector& t_BusIntrospector)
{
	m_DeviceAboutAnnouncement = &t_DeviceAboutAnnouncement;
	m_BusIntrospector = &t_BusIntrospector;

	m_ReadingNotifications = true;
}

void NotificationValidator::Receive(ajn::services::Notification const& t_Notification)
{
	if (m_ReadingNotifications)
	{
		m_NotificationHandler->Receive(t_Notification);

		readNotification();
	}
}

void NotificationValidator::Dismiss(const int32_t arg0, const qcc::String arg1) {}

void NotificationValidator::readNotification()
{
	ajn::services::Notification* notification;

	while ((notification = m_NotificationHandler->getReceivedNotification()) != nullptr)
	{
		LOG(INFO) << "Validating notification from DUT with messageId: "
			<< notification->getMessageId();

		m_NotificationCounter++;

		bool includedRichIconUrl = false;
		bool includedRichAudioUrl = false;
		bool includedResponseObjectPath = false;

		LOG(INFO) << "Checking Version field of received notification message";
		EXPECT_EQ(notification->getVersion(), m_ExpectedNotificationVersion)
			<< "Notification version received (" << notification->getVersion()
			<< ") does not match IXITN_NotificationVersion ("
			<< m_ExpectedNotificationVersion << ")";

		LOG(INFO) << "Checking AppName field of received notification message";
		EXPECT_EQ(std::string(notification->getAppName()), std::string(m_DeviceAboutAnnouncement->getAppName()))
			<< "AppName parameter received in notification: " << notification->getAppName()
			<< " does not match About Announcement parameter: "
			<< m_DeviceAboutAnnouncement->getAppName();

		LOG(INFO) << "Checking DeviceName field of received notification message";
		EXPECT_EQ(std::string(notification->getDeviceName()), std::string(m_DeviceAboutAnnouncement->getDeviceName()))
			<< "DeviceName parameter received in notification: " << notification->getDeviceName()
			<< " does not match About Announcement parameter: "
			<< m_DeviceAboutAnnouncement->getDeviceName();

		LOG(INFO) << "Checking notification message type";
		EXPECT_NE(notification->getMessageType(), ajn::services::NotificationMessageType::UNSET);

		const char* richIconUrl = notification->getRichIconUrl();
		if (m_ExpectedRichIconUrl && richIconUrl != nullptr)
		{
			validateUrl(richIconUrl);
		}
		else if ((m_ExpectedRichIconUrl && richIconUrl == nullptr)
			|| (!m_ExpectedRichIconUrl && richIconUrl != nullptr))
		{
			FAIL() << "RichIconUrl support is not the defined by ICSN_RichIconUrl";
		}
		
		const std::vector<ajn::services::RichAudioUrl> richAudioUrls = notification->getRichAudioUrl();
		if (m_ExpectedRichAudioUrl && !richAudioUrls.empty())
		{
			for (auto richAudioUrl : richAudioUrls)
			{
				const qcc::String url_string = richAudioUrl.getUrl();
				validateUrl(url_string.c_str());
			}
		}
		else if ((m_ExpectedRichAudioUrl && richAudioUrls.empty())
			|| (!m_ExpectedRichAudioUrl && !richAudioUrls.empty()))
		{
			FAIL() << "RichAudioUrl support is not the defined by ICSN_RichAudioUrl";
		}

		const char* responseObjectPath = notification->getControlPanelServiceObjectPath();
		if (m_ExpectedResponseObjectPath && responseObjectPath != nullptr)
		{
			LOG(INFO) << "Control Panel Service response object path: " << responseObjectPath;

			m_BusIntrospector->introspect(responseObjectPath);
		}
		else if ((m_ExpectedResponseObjectPath && responseObjectPath == nullptr)
			|| (!m_ExpectedResponseObjectPath && responseObjectPath != nullptr))
		{
			FAIL() << "RespObjectPath support is not the defined by ICSN_RespObjectPath";
		}

		std::string message("Notification Received: ");
		message.append("messageType: ");
		message.append(MessageSet("", notification->getMessageType()).getPriorityAsString());
		message.append("; richIconUrl: ");
		message.append(includedRichIconUrl ? "true" : "false");
		message.append("; richAudioUrl: ");
		message.append(includedRichAudioUrl ? "true" : "false");
		message.append("; responseObject: ");
		message.append(includedResponseObjectPath ? "true" : "false");
		LOG(INFO) << message;
	}
}

unsigned int NotificationValidator::getNumberOfNotificationsReceived()
{
	return static_cast<unsigned int>(m_NotificationCounter);
}

void NotificationValidator::validateUrl(const std::string& t_Url)
{
	std::regex urlRegex(URL_REGEX, std::regex::extended);
	std::smatch urlMatchResult;

	ASSERT_TRUE(std::regex_match(t_Url, urlMatchResult, urlRegex))
		<< "Malformed url.";
	SUCCEED() << "Url format is correct";
}

